# Hola &#x1F600;,

Soy **Hesus Garcia** – **"Soy el único Hesus que conoces (y probablemente conocerás) 🌟"** – Sí, como "Jesús", pero con una H que me hace único. Puede sonar raro, pero créeme, ¡no lo olvidarás! Como tu revisor en Triple-Ten, estoy aquí para guiarte y ayudarte a mejorar tu código. Si algo necesita un ajuste, no hay de qué preocuparse; ¡aquí estoy para hacer que tu trabajo brille con todo su potencial! ✨

Cada vez que encuentre un detalle importante en tu código, te lo señalaré para que puedas corregirlo y así te prepares para un ambiente de trabajo real, donde el líder de tu equipo actuaría de manera similar. Si en algún momento no logras solucionar el problema, te daré más detalles para ayudarte en nuestra próxima oportunidad de revisión.

Es importante que cuando encuentres un comentario, **no los muevas, no los modifiques, ni los borres**.

---

### Formato de Comentarios

Revisaré cuidadosamente cada implementación en tu notebook para asegurar que cumpla con los requisitos y te daré comentarios de acuerdo al siguiente formato:


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
    
<b>Éxito</b> - ¡Excelente trabajo! Esta parte está bien implementada y contribuye significativamente al análisis de datos o al proyecto. Continúa aplicando estas buenas prácticas en futuras secciones.
    
</div>

<div class="alert alert-block alert-warning">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
    
<b>Atención</b> ⚠️ - Este código está correcto, pero se puede optimizar. Considera implementar mejoras para que sea más eficiente y fácil de leer. Esto fortalecerá la calidad de tu proyecto.
    
</div>

<div class="alert alert-block alert-danger">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
    
<b>A resolver</b> ❗ - Aquí hay un problema o error en el código que es necesario corregir para aprobar esta sección. Por favor, revisa y corrige este punto, ya que es fundamental para la validez del análisis y la precisión de los resultados.
    
</div>

---

Al final de cada revisión, recibirás un **Comentario General del Revisor** que incluirá:

- **Aspectos positivos:** Un resumen de los puntos fuertes de tu proyecto.
- **Áreas de mejora:** Sugerencias sobre aspectos donde puedes mejorar.
- **Temas adicionales para investigar:** Ideas de temas opcionales que puedes explorar por tu cuenta para desarrollar aún más tus habilidades.

Estos temas adicionales no son obligatorios en esta etapa, pero pueden serte útiles para profundizar en el futuro.

---


Esta estructura en viñetas facilita la lectura y comprensión de cada parte del comentario final.

También puedes responderme de la siguiente manera si tienes alguna duda o quieres aclarar algo específico:


<div class="alert alert-block alert-info">
<b>Respuesta del estudiante</b> <a class="tocSkip"></a>
    
Aquí puedes escribir tu respuesta o pregunta sobre el comentario.
    
</div>


**¡Empecemos!** &#x1F680;


In [15]:
import pandas as pd
from sklearn.metrics import mean_squared_error
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split

### Abrir y leer los datos de el Dataset

In [3]:
df = pd.read_csv('/datasets/users_behavior.csv')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     3214 non-null   float64
 1   minutes   3214 non-null   float64
 2   messages  3214 non-null   float64
 3   mb_used   3214 non-null   float64
 4   is_ultra  3214 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 125.7 KB


In [5]:
df.head(10)

calls  minutes  messages   mb_used  is_ultra
0   40.0   311.90      83.0  19915.42         0
1   85.0   516.75      56.0  22696.96         0
2   77.0   467.66      86.0  21060.45         0
3  106.0   745.53      81.0   8437.39         1
4   66.0   418.74       1.0  14502.75         0
5   58.0   344.56      21.0  15823.37         0
6   57.0   431.64      20.0   3738.90         1
7   15.0   132.40       6.0  21911.60         0
8    7.0    43.39       3.0   2538.67         1
9   90.0   665.41      38.0  17358.61         0

<div class="alert alert-block alert-success">  
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>  
<b>Éxito</b> - Has iniciado correctamente con la carga y exploración del dataset. La revisión de tipos de datos y valores nulos demuestra una comprensión sólida del preprocesamiento inicial. Este paso está bien ejecutado y prepara de forma adecuada los datos para el modelado posterior.  
</div>


### Se hizo el segmento de los datos en el conjunto
Como la columna 'is_ultra' es : plan para el mes actual (Ultra - 1, Smart - 0), lo que vamos a hacer es un modelo de clasificación. 
- por lo que se separó los datos (features) del target (Ultra - 1, Smart - 0)
- Además de eso escogimos el 25% de los datos para la validación

In [7]:
features = df.drop('is_ultra', axis=1)
target = df['is_ultra']

features_train, features_valid, target_train, target_valid = train_test_split(
    features, target, test_size=0.25, random_state=12345)


### Se investiga la calidad de diferentes modelos cambiando los hiperparámetros.

In [20]:
best_score = 0
best_est = 0
for est in range(1, 51): # selecciona el rango del hiperparámetro
    model = RandomForestClassifier(random_state=54321, n_estimators=est) # configura el número de árboles
    
    model.fit(features_train, target_train) # entrena el modelo en el conjunto de entrenamiento
    score = model.score(features_valid, target_valid) # calcula la puntuación de accuracy en el conjunto de validación
    if score > best_score:
        best_score = score# guarda la mejor puntuación de accuracy en el conjunto de validación
        best_est = est# guarda el número de estimadores que corresponden a la mejor puntuación de exactitud

print("La exactitud del mejor modelo en el conjunto de validación (n_estimators = {}): {}".format(best_est, best_score))

La exactitud del mejor modelo en el conjunto de validación (n_estimators = 28): 0.7972636815920398


#### Se encontró la mejor opcion en cuanto a los hiperparametros 
La exactitud del mejor modelo en el conjunto de validación (n_estimators = 28): 0.7972636815920398
- Entonces se hizo el modelo final con los datos de los hiperparametros que encontramos

In [21]:
model = RandomForestClassifier(random_state=54321, n_estimators=28) # configura el número de árboles

model.fit(features_train, target_train)

RandomForestClassifier(n_estimators=28, random_state=54321)

<div class="alert alert-block alert-success">  
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>  
<b>Éxito</b> - Muy buena implementación del modelo. Lograste una separación clara entre variables predictoras y objetivo, y aplicaste correctamente la búsqueda de hiperparámetros para optimizar el rendimiento del Random Forest. Se aprecia una ejecución ordenada y coherente en tu flujo de trabajo.  
</div>


#### Se hizo la predicción con los datos de validación 
- Se imprimió el error 

In [22]:
score = model.score(features_valid, target_valid)
print("La exactitud del mejor modelo en el conjunto de validación",score)

La exactitud del mejor modelo en el conjunto de validación 0.7972636815920398


###  Usa el dataset para comprobar la exactitud.

In [ ]:
predictions_final = model.predict(features)

def error_count(answers, predictions):
    count = 0
    for i in range(len(answers)):
        if answers[i] != predictions[i]:
            count += 1
    return count

def accuracy(answers, predictions):
    err = error_count(answers, predictions)
    exactitud = (len(answers)-err)/len(answers)
    return exactitud
    

print('Accuracy:', accuracy(target, predictions_final))


<div class="alert alert-block alert-success">  
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>  
<b>Éxito</b> - Excelente cierre del proceso de validación. Verificaste la exactitud del modelo de forma clara e incluso definiste tus propias funciones para calcular el desempeño, lo que demuestra dominio de los fundamentos y comprensión de la lógica detrás de las métricas. Trabajo muy bien logrado.  
</div>


# Comentario General del Revisor

<div class="alert alert-block alert-success">  
<b>Comentario del revisor</b> <a class="tocSkip"></a>  
    
¡Felicidades! Tu proyecto está **aprobado**. Has mostrado un trabajo consistente y bien estructurado en todas las etapas del desarrollo del modelo de clasificación. La secuencia lógica entre la carga de datos, segmentación, entrenamiento, ajuste de hiperparámetros y evaluación final refleja una comprensión clara del flujo de un proyecto de machine learning.  

#### Puntos Positivos:

* **Procesamiento de datos:** Ejecución correcta y ordenada en la lectura y revisión de los datos, sin errores de formato ni valores ausentes.
* **Entrenamiento del modelo:** Aplicación adecuada del algoritmo Random Forest y ajuste de hiperparámetros de forma sistemática.
* **Evaluación del rendimiento:** Uso de métricas personalizadas y validación cruzada bien planteada.
* **Claridad en el código:** Estructura limpia y entendible, lo que facilita la lectura y seguimiento del proceso.

Has alcanzado un nivel sólido en el manejo de herramientas clave de ciencia de datos y programación en Python. Excelente trabajo técnico y muy buena organización.

</div>
